### PR TITLE
Correct CMake issues

### PR DESCRIPTION
**Pull request by doodspav**
- install properly without forcing us upon our consumers' consumers
- remove unnecessary options (such as platform specific ones, or ones that have no impact on building/installing)
- support alias targets helping consumers catch errors sooner
- make CMake files more structured and readable
- change minimum required version from `3.8 .. 3.23` to just `3.14`
- all symbols are now private (you should probably address this asap)
- more stuff properly, i don't remember

### CODE REVIEW

**Comment by jeremy-rifkin** [comment]
Thank you for the PR! I will review as soon as I can.

**Comment by doodspav** [comment]
Follow up things you should probably do:
- check your flags; i removed a whole bunch because they don't affect the build (they're compiler warnings) and they belong in presets, not in the CML file
- check your macro definitions; i removed some that set some macros that didn't affect build/install (they should stay removed, but you might want to update your docs concerning that)
- fix the CMake for your tests, and maybe use a proper test framework
- fix the CMake for your dependencies (like `cpptrace`) which gets installed along with `libassert` even when being linked as a private dependency
- you now have a new header file, `<assert/assert_export.hpp>` that gets generated by CMake; you can include this without needing a relative include path, and you should mark all the symbols you want to be exported with `ASSERT_EXPORT` because right now all symbols are private (even on linux)

**Comment by doodspav** [comment]
Bear in mind that `cpptrace`'s CMake breaks consuming an installed version of `libassert` consumed via `find_package`

**Comment by jeremy-rifkin** [comment]
Thank you again for this PR, I have integrated much of it into cpptrace on the dev branch and things seem to work well there. I will give this PR a more thorough review this weekend.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,230 +1,164 @@
...
+    # src
+    src/assert.cpp
 )
 
-target_compile_options(
-  assert
-  PRIVATE
-  $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wall -Wextra -Werror=return-type -Wshadow -Wundef>
-  $<$<CXX_COMPILER_ID:GNU>:-Wuseless-cast -Wnonnull-compare>
-  $<$<CXX_COMPILER_ID:MSVC>:/W4 /WX /permissive->
```
Comment (jeremy-rifkin) [inline CMakeLists.txt:69]: I am likely going to re-add these warnings, they're important for development

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,230 +1,164 @@
...
-    )
-    add_test(
-      NAME integration
-      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/tests
-      COMMAND
-      python3 run-tests.py $<TARGET_FILE:integration> ${CMAKE_BUILD_TYPE} ${CMAKE_CXX_COMPILER_ID} ${CMAKE_CXX_STANDARD}
-    )
+# ---- Setup Tests ----
+
+if(ASSERT_BUILD_TESTING)
```
Comment (jeremy-rifkin) [inline CMakeLists.txt:148]: If this flag is being changed the ci scripts will need to be updated. I can take care of that.

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            add_executable(${test_name} ${test_file})
+            target_link_libraries(${test_name} libassert-lib)
+            add_test(NAME ${test_name} COMMAND ${test_name})
+        endforeach(test_file ${unit_test_sources})
+    # endif()
+    # if(ASSERT_BUILD_BASIC_TEST)
+        add_executable(basic tests/basic/basic_test.cpp)
+        target_link_libraries(basic PRIVATE libassert-lib)
+    # endif()
+endif()
```
Comment (jeremy-rifkin) [inline tests/CMakeLists.txt:68]: A newline at the end of this file would be good

**Review by jeremy-rifkin — APPROVED** [review]
I think this looks really good overall. I commented on some super minor things I noticed, I can take care of them after I merge.

**Pull request merged**